### PR TITLE
More context variables for error handlers, see CORE-1132 and CORE-2040.

### PR DESCRIPTION
**Pull request by dyemanov**
I'm proposing it via a pull request to get some comments. Implementation is pretty straightforward, but I'm somewhat worried about the keywords. EXCEPTION seems to be a logical counterpart for GDSCODE/SQLCODE/SQLSTATE, using the same keywords as in handlers: WHEN {GDSCODE | SQLCODE | SQLSTATE | EXCEPTION | ANY} DO. ERROR_MESSAGE is taken from MSSQL, Oracle uses a shorter abbreviation ERRMSG. I failed to find anything like that in MySQL or PostgreSQL. A separate question is whether we should also implement a context variable for "exception message" raised using isc_except (mentioned in CORE-2040).

### CODE REVIEW

File: src/jrd/constants.h [hunk 1]
```diff
@@ -452,4 +454,7 @@ const int OPT_STATIC_ITEMS = 64;
 const int WITH_GRANT_OPTION = 1;
 const int WITH_ADMIN_OPTION = 2;
 
+// Max length of the string returned by ERROR_TEXT context variable
+const USHORT MAX_ERROR_MSG_LENGTH = 4096; // corresponds to 1024 UTF-8 characters
```
Comment (asfernandes) [inline src/jrd/constants.h]: It's better to use 1024 * METADATA_BYTES_PER_CHAR here.

Comment (dyemanov) [inline src/jrd/constants.h]: Done, thanks.

**Comment by sim1984** [comment]
Is not there a conflict between re-raising an exception statement and the context variable for the name of an exception?

Example 1:
```
BEGIN
      ...
   WHEN ANY DO
     IF (EXCEPTION IS NOT NULL)
        EXECUTE PROCEDURE P_USR_EXCEPTION(EXCEPTION);
     ELSE
        EXECUTE PROCEDURE P_SYS_EXCEPTION;
END
```

Example 2:

```
CREATE PROCEDURE ADD_COUNTRY (
    ACountryName COUNTRYNAME,
    ACurrency VARCHAR(10) )
AS
BEGIN
  INSERT INTO country (country,
                       currency)
  VALUES (:ACountryName,
          :ACurrency);
  WHEN ANY DO
  BEGIN
    IN AUTONOMOUS TRANSACTION DO
      INSERT INTO ERROR_LOG (PSQL_MODULE,
                             GDS_CODE,
                             SQL_CODE,
                             SQL_STATE,
                             EXCEPTION_NAME,
                             ERR_MSG)
      VALUES ('ADD_COUNTRY',
              GDSCODE,
              SQLCODE,
              SQLSTATE,
              EXCEPTION,
              ERROR_MESSAGE);
    -- re-raise exception
    EXCEPTION;
  END
END 
```

**Comment by dyemanov** [comment]
Not really, although it might look confusing. EXCEPTION as value and EXCEPTION as statement cannot be intermixed in any context.

**Comment by FbnDev** [comment]
```
WHEN ANY E DO
BEGIN
    IN AUTONOMOUS TRANSACTION DO
      INSERT INTO ERROR_LOG (
                             SQL,
                             PLACE,
                             TRACE,
                             GDS_CODE,
                             SQL_CODE,
                             SQL_STATE,
                             EXCEPTION_NAME,
                             ERR_MSG)
      VALUES (
              E.SQL,
              E.LINE || ':' || E.POSITION,
              E.TRACE,
              E.GDSCODE,
              E.SQLCODE,
              E.SQLSTATE,
              E.NAME,
              E.MESSAGE);
    -- re-raise exception
    RAISE;
END
```
and
```
WHEN ANY E DO
BEGIN
  LOG_EXCEPTION(E);
  -- LOG_EXCEPTION declared as LOG_EXCEPTION(RDB$E_EXCEPTION E)
  RAISE;
END
```
and
```
  DECLARE EXCEPTION E;
BEGIN
.........
  IF SOME_CONDITION THEN
  BEGIN
    E = CREATE_EXCEPTION(......);
    RAISE E;
  END
END
```
and
```
WHEN RDB$E_FOREIGN_KEY_VIOLATION E DO
BEGIN
  LOG_FK_VIOLATION(E.CONSTRAINT_NAME, E.KEY_VALUE);
  RAISE;
END 
ELSE WHEN RDB$E_USER E DO
BEGIN
  IF (E.EXCEPTION = SOME_USER_EXCEPTION) THEN
  BEGIN
    LOG_USER_EXCEPTION(E);
    -- LOG_USER_EXCEPTION declared as LOG_USER_EXCEPTION(RDB$E_USER E)
    -- no re-rise
  END
  ELSE 
  BEGIN
    LOG_OTHER_USER_EXCEPTION(E);
    -- LOG_OTHER_USER_EXCEPTION declared as LOG_OTHER_USER_EXCEPTION(RDB$E_USER E)
    RAISE;
  END 
END 
ELSE WHEN ANY E DO
BEGIN
  IF (E IS RDB$E_PRIMARY_KEY_VIOLATION) THEN
  BEGIN 
    LOG_PK_VIOLATION(CAST(E AS RDB$E_PRIMARY_KEY_VIOLATION));
    -- or unsafe cast: LOG_PK_VIOLATION(E); 
    -- LOG_PK_VIOLATION declared as LOG_PK_VIOLATION(RDB$E_PRIMARY_KEY_VIOLATION E)
    RAISE;
  END
  ELSE
  BEGIN 
    LOG_EXCEPTION(E);
    -- LOG_EXCEPTION declared as LOG_EXCEPTION(RDB$E_EXCEPTION E)
    RAISE;
  END
END
```

**Pull request merged**